server/snapshot.go

```diff
@@ -17,13 +17,131 @@ import (
 const snapshotBatchSize = 200
 
 var (
-	fragmentPool     = &sync.Pool{New: func() interface{} { return &spb.RaftSnapshotFragment{} }}
-	batchPool        = &sync.Pool{New: func() interface{} { return &spb.Batch{} }}
-	msgsSnapshotPool = &sync.Pool{New: func() interface{} { return &spb.MessagesSnapshot{} }}
-	subSnapshotPool  = &sync.Pool{New: func() interface{} { return &spb.SubSnapshot{} }}
+	fragmentPool       = &sync.Pool{New: func() interface{} { return &spb.RaftSnapshotFragment{} }}
+	batchPool          = &sync.Pool{New: func() interface{} { return &spb.Batch{} }}
+	msgsSnapshotPool   = &sync.Pool{New: func() interface{} { return &spb.MessagesSnapshot{} }}
+	subSnapshotPool    = &sync.Pool{New: func() interface{} { return &spb.SubSnapshot{} }}
+	clientSnapshotPool = &sync.Pool{New: func() interface{} { return &spb.ClientSnapshot{} }}
 )
 
-// channelSnapshot implements the raft.FSMSnapshot interface.
+// serverSnapshot implements the raft.FSMSnapshot interface by snapshotting
+// StanServer state.
+type serverSnapshot struct {
+	*StanServer
+}
+
+func newServerSnapshot(s *StanServer) raft.FSMSnapshot {
+	return &serverSnapshot{s}
+}
+
+// Persist should dump all necessary state to the WriteCloser 'sink',
+// and call sink.Close() when finished or call sink.Cancel() on error.
+func (s *serverSnapshot) Persist(sink raft.SnapshotSink) (err error) {
+	defer func() {
+		if err != nil {
+			sink.Cancel()
+		}
+	}()
+
+	// Snapshot clients. We are only snapshotting the client metadata here
+	// (heartbeat inbox and client ID). Subscriptions are replicated through
+	// the channel Raft groups, so upon snapshot restore, the server will
+	// load any of the client's subscriptions from the store.
+	var buf [4]byte
+	for _, client := range s.clients.getClients() {
+		fragment := newFragment()
+		fragment.FragmentType = spb.RaftSnapshotFragment_Conn
+		clientSnap := clientSnapshotPool.Get().(*spb.ClientSnapshot)
+		client.RLock()
+		clientSnap.Info = &client.info.ClientInfo
```

Comment:
I believe that the client's info is immutable so we should be ok here not making a copy of ClientInfo.

Reply:
Is this possible with protobuf? I think it requires a pointer.

Reply:
Yes, since you use an embedded message, it is stored as pointer, what I meant is that you could do this:
```
copyInfo := client.info.ClientInfo
clientSnap.Info = &copyInfo
```
In that way, if client.info.ClientInfo struct content were to change, clientSnap.Info would not be affected. But again, I believe that this is immutable so we should be safe.

Reply:
Ah yes, I misunderstood your original comment. I will leave this as is.